### PR TITLE
K.krupa/raytracing texture

### PR DESCRIPTION
**Pull request by karolk98**

### CODE REVIEW

File: RayTracer/Source/Materials/Diffuse.cs [hunk 1]
```diff
@@ -6,19 +6,25 @@ namespace RayTracing.Materials
...
+        {
+            _sampler = sampler ?? new ThreadSafeSampler<Vector3>(Vec3Sampling.UniformSphere, 10000, 8, Vec3Sampling.ToSphereSurface);
+            Albedo = new SolidColor(albedo);
+        }
 
         public bool Scatter(ref Ray ray, ref HitInfo hit, out Color attenuation, out Ray scattered)
         {
             scattered = new Ray(hit.HitPoint, hit.Normal + _sampler.Sample);
-            attenuation = Albedo;
+            attenuation = Albedo[hit.TexCoord.X,hit.TexCoord.Y];
```
Comment (p-flis) [inline RayTracer/Source/Materials/Diffuse.cs:27]: add space

File: RayTracer/Source/Materials/Reflective.cs [hunk 2]
```diff
@@ -21,7 +29,7 @@ public bool Scatter(ref Ray ray, ref HitInfo hit, out Color attenuation, out Ray
         {
             Vector3 reflected = ray.Direction.Reflect(hit.Normal);
             scattered = new Ray(hit.HitPoint, reflected + Disturbance * _sampler.Sample);
-            attenuation = Albedo;
+            attenuation = Albedo[hit.TexCoord.X,hit.TexCoord.Y];
```
Comment (p-flis) [inline RayTracer/Source/Materials/Reflective.cs:32]: space

File: RayTracerApp/Forms/MainForm.cs [hunk 3]
```diff
@@ -71,13 +77,15 @@ private void GLControl_Load(object sender, EventArgs e)
...
-                Material = new Diffuse(Color.FromColor4(Color4.ForestGreen))
+                Material = new Reflective(new Texture("earthmap.jpg"), 0.75f)
             }.Load());
+            _scene.AddModel(new Cylinder(2) {Position = new Vector3(5f, 0.5f, 0), Scale = 1, Material = new Diffuse(Color.FromColor4(Color4.Chocolate))});
+            _scene.AddModel(plane);
+            // _scene.AddModel(new Sphere
+            // {
+            //     Position = new Vector3(0, 0, 0), Scale = 20, 
+            //     Material = new Diffuse(Color.FromColor4(Color4.DarkBlue))
+            // }.Load());
```
Comment (p-flis) [inline RayTracerApp/Forms/MainForm.cs:88]: delete commented code

**Review by p-flis — APPROVED** [review]

**Review by tomasz-herman — COMMENTED** [review]
Please add textured something to demo project as well

File: RayTracer/Source/Materials/Reflective.cs [hunk 4]
```diff
@@ -1,16 +1,24 @@
...
     public class Reflective : IMaterial
     {
-        public Color Albedo { get; set; }
+        public ITexture Albedo { get; set; }
         public float Disturbance { get; set; }
         private readonly AbstractSampler<Vector3> _sampler;
 
         public Reflective(Color albedo, float disturbance = 0, AbstractSampler<Vector3> sampler = null)
+        {
+            _sampler = sampler ?? new ThreadSafeSampler<Vector3>(Vec3Sampling.UniformSphere, 125, 8);
```
Comment (tomasz-herman) [inline RayTracer/Source/Materials/Reflective.cs:16]: Please increase samples here to 10k as well it looks blocky  
Comparison  
sample

Comment (tomasz-herman) [reply RayTracer/Source/Materials/Reflective.cs:16]: 10k vs 125

File: RayTracer/Source/Materials/Reflective.cs [hunk 5]
```diff
@@ -1,16 +1,24 @@
...
+            Albedo = new SolidColor(albedo);
+            Disturbance = disturbance;
+        }
+
+        public Reflective(ITexture albedo, float disturbance = 0, AbstractSampler<Vector3> sampler = null)
         {
             _sampler = sampler ?? new ThreadSafeSampler<Vector3>(Vec3Sampling.UniformSphere, 125, 8);
```
Comment (tomasz-herman) [inline RayTracer/Source/Materials/Reflective.cs:23]: 10k here as well

**Review by tomasz-herman — APPROVED** [review]

**Pull request merged**